flags/test.ts

```diff
@@ -736,3 +737,958 @@ Deno.test("valueFollowingDoubleHyphenIsNotUnknown", function (): void {
+            bab: boolean;
+          };
+        };
+        _: Array<string | number>;
+      }
+    >
+  >(true);
+});
+
+Deno.test("typesOfArgsWithDottedDefaults", function (): void {
+  const argv = parse([], {
+    string: ["foo"],
+    default: {
+      "foo.bar": 1,
+    },
+  });
+  assertType<
+    IsExact<
+      typeof argv,
+      & { [x: string]: unknown }
+      & {
+        foo: string | {
+          bar: number;
+        };
+        _: Array<string | number>;
+      }
+    >
+  >(true);
+});
+
+/** ------------------------ COLLECT OPTION -------------------------- */
+
+Deno.test("typesOfCollectAllArgs", function (): void {
+  const argv = parse([], {
+    boolean: ["foo", "dotted.beep"],
+    string: ["bar", "dotted.boop"],
+    collect: true,
+  });
+  assertType<
+    IsExact<
+      typeof argv,
+      & { [x: string]: Array<unknown> }
+      & {
+        bar?: Array<string> | undefined;
+        dotted: {
+          boop?: Array<string> | undefined;
+          beep: Array<boolean>;
+        };
+        foo: Array<boolean>;
+        _: Array<string | number>;
+      }
+    >
+  >(true);
+});
+
+Deno.test("typesOfCollectAllArgsWithDefaults", function (): void {
+  const argv = parse([], {
+    boolean: ["foo", "dotted.beep"],
+    string: ["bar", "dotted.boop"],
+    collect: true,
+    default: {
+      bar: 123,
+      dotted: {
+        beep: new Date(),
+        boop: /.*/,
+      },
+      foo: new TextDecoder(),
+    },
+  });
+  assertType<
+    IsExact<
+      typeof argv,
+      & { [x: string]: Array<unknown> }
+      & {
+        bar: number | Array<string>;
+        foo: TextDecoder | Array<boolean>;
+        dotted: {
+          beep: Array<boolean> | Date;
+          boop: RegExp | Array<string>;
+        };
+        _: Array<string | number>;
+      }
+    >
+  >(true);
+});
+
+Deno.test("typesOfCollectArgs", function (): void {
+  const argv = parse([], {
+    boolean: ["foo", "dotted.beep"],
+    string: ["bar", "dotted.boop"],
+    collect: ["foo", "dotted.boop"],
+  });
+  assertType<
+    IsExact<
+      typeof argv,
+      & { [x: string]: unknown }
+      & {
+        bar?: string | undefined;
+        dotted: {
+          boop?: Array<string> | undefined;
+          beep: boolean;
+        };
+        foo: Array<boolean>;
+        _: Array<string | number>;
+      }
+    >
+  >(true);
+});
+
+Deno.test("typesOfCollectArgsWithDefaults", function (): void {
+  const argv = parse([], {
+    boolean: ["foo", "dotted.beep"],
+    string: ["bar", "dotted.boop"],
+    collect: ["foo", "dotted.boop"],
+    default: {
+      bar: 123,
+      dotted: {
+        beep: new Date(),
+        boop: /.*/,
+      },
+      foo: new TextDecoder(),
+    },
+  });
+  assertType<
+    IsExact<
+      typeof argv,
+      & { [x: string]: unknown }
+      & {
+        bar: number | string;
+        foo: TextDecoder | Array<boolean>;
+        dotted: {
+          beep: boolean | Date;
+          boop: RegExp | Array<string>;
```

Comment:
This boop seems possible to be `string` when single `--dotted.boop foo` is passed

Reply:
Ah, this might depend on #2247?

Reply:
Yep, it depends on the other PR. Collectable options will then always be an array.